src/contracts/GPv2Settlement.sol

```diff
@@ -300,6 +300,23 @@ contract GPv2Settlement {
         filledAmount[uidKey] = currentFilledAmount;
     }
 
+    /// @dev Allows settlment function to make arbitrary contract executions.
+    /// @param interaction contains address and calldata of the contract interaction.
+    function executeInteraction(GPv2Encoding.Interaction memory interaction)
+        internal
+    {
+        // To prevent possible attack on user funds, we explicitly disable
+        // interactions with AllowanceManager contract.
+        require(
+            interaction.target != address(allowanceManager),
+            "GPv2: forbidden interaction"
+        );
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory response) =
+            (interaction.target).call(interaction.callData);
+        require(success, string(response));
```

Comment:
You could do potentially do something like `string(abi.encodePacked("GPv2 Interaction: ", response))` to make the error message even more awesome.

Reply:
Cool idea! In that case I would recommend
```
if (!success) {
  revert(string(abi.encodePacked(...));
}
```
Just so we don't build a string on success as well.

It would be worth testing to make sure that this makes a difference though.

Reply:
Right, this is great. SO I have made this suggestion and plan to keep it this way, but we see unexpected error messages with this packed encoding and also with just regular encoding:

```
AssertionError: Expected transaction to be reverted with GPv2 Interaction: test error, but other exception was thrown: Error: VM Exception while processing transaction: revert GPv2 Interaction:�y� 
test error
```
However, we can assert that the error string is just "test error" and everything passes. I hope this is alright for everyone.

Reply:
I think its because you need to strip the `bytes4(keccak256("Error(string)"))` from the reverted bytes. I really like the suggestion of concating the error messages to produce a better one.

Reply:
I am not sure I follow this suggestion... Should I introduce some sort of string slicing method to deal with this?

Reply:
I could use 

https://github.com/GNSPS/solidity-bytes-utils

Reply:
> Should I introduce some sort of string slicing method to deal with this?

I think Solidity has built-in slicing `response[4:]`

Reply:
WHAT! This must be new! Because it never had it before.

Reply:
My point is that `require(success, string(response));` is also wrong. I just tried it out and the actual revert message you get is:
```
     Error: VM Exception while processing transaction: revert ^H�y�^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@ ^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
test error^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
```

(Here is the diff from the current `HEAD`:
```diff
git diff
diff --git a/src/contracts/GPv2Settlement.sol b/src/contracts/GPv2Settlement.sol
index 6ff90d0..25049de 100644
--- a/src/contracts/GPv2Settlement.sol
+++ b/src/contracts/GPv2Settlement.sol
@@ -315,9 +315,7 @@ contract GPv2Settlement {
         (bool success, bytes memory response) =
             (interaction.target).call(interaction.callData);
         // solhint-enable avoid-low-level-calls
-        if (!success) {
-            revert(string(abi.encodePacked("GPv2 Interaction: ", response)));
-        }
+        require(success, string(response));
     }
 
     /// @dev Extracts specific order information from the standardized unique
diff --git a/test/GPv2Settlement.test.ts b/test/GPv2Settlement.test.ts
index 517a073..91928f9 100644
--- a/test/GPv2Settlement.test.ts
+++ b/test/GPv2Settlement.test.ts
@@ -695,7 +695,7 @@ describe("GPv2Settlement", () => {
       ).to.be.revertedWith("GPv2: forbidden interaction");
     });
 
-    it("should fail when interaction reverts", async () => {
+    it.only("should fail when interaction reverts", async () => {
       const reverter = await waffle.deployMockContract(deployer, [
         "function alwaysReverts()",
       ]);
@@ -709,6 +709,7 @@ describe("GPv2Settlement", () => {
         callData: revertingCallData,
       };
 
+      await settlement.callStatic.executeInteractionTest(failingInteraction);
       await expect(
         settlement.callStatic.executeInteractionTest(failingInteraction),
       ).to.be.revertedWith("test error");
```

Reply:
TypeError: Index range access is only supported for dynamic calldata arrays.

Reply:
`^H�y�` is how my terminal prints out `0x08c379a0` (which is the "error selector" and has the value of `bytes4(keccak256("Error(string)"))`.

Reply:
> TypeError: Index range access is only supported for dynamic calldata arrays.

Damn, so this will be a bit harder.

Reply:
You might have to use assembly like the [`StorageAccessible` contract](https://github.com/gnosis/util-contracts/blob/1d84e3d78fd8eda73cd48c0cde53fb70a97fbece/contracts/StorageAccessible.sol#L86-L88):
```sol
        assembly {
            revert(add(response, 0x20), mload(response))
        }
```

This is without string concatenation though (which would require some additional code)

Reply:
> I could use
>
> https://github.com/GNSPS/solidity-bytes-utils

Personally, I don't think its worth adding a dependency for this, but I don't know how the others feel. Also considering the `BytesLib` is `pragma solidity >=0.5.0 <0.7.0;` I think it would be more work (it doesn't support the Solidity version used here).

Reply:
Ok, so I have now managed to slice the first 4 bytes off the response (with a loop)

```
            bytes memory message = new bytes(response.length - 4);
            for (uint256 i = 4; i < response.length; i++) {
                message[i - 4] = response[i];
            }
```

but still getting a line break in the revert message...

Reply:
What's even stranger, is that if I replace the value 4 in the above snippet with 5 or even 6... the revert message is the same.

Reply:
Perhaps we can just revert with the response and deal with this last detail of concatenation in a separate issue.

Reply:
I have created this issue #240 and would propose falling back on 

```
require(success, string(response));
```
for the time being.

Reply:
`require(success, string(response));` does not work, see my comment above: https://github.com/gnosis/gp-v2-contracts/pull/224#discussion_r539206903

Reply:
But it does work!
```
await expect(
        settlement.callStatic.executeInteractionTest(failingInteraction),
      ).to.be.revertedWith("test error");
```
Passes.
```
  GPv2Settlement
    executeInteraction
      ✓ should fail when interaction reverts (301ms)
```

Reply:
In any case, I can also revert with `"GPv2: Failed Interaction"`

Reply:
> The test passes because test error is included in the revert string, but it also includes a bunch of other random bytes that shouldn't be there.

I see, then why not just leave it with the concatenation which has some line breaks, but also passes with expected message of "test error"

Reply:
> But it does work! [...] Passes

The test passes but the revert message is
```
     Error: VM Exception while processing transaction: revert ^H�y�^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@ ^@^@^@^@^@
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
test error^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@
```

Reply:
Settled on
```js
        if (!success) {
            assembly {
                revert(add(response, 0x20), mload(response))
            }
        }
```

With an open TODO (#240)